src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```diff
@@ -197,439 +107,104 @@ public String version() {
-                      ? changeStreamDocument
-                      : new BsonDocument(ID_FIELD, changeStreamDocument.get(ID_FIELD));
-
-              createSourceRecord(
-                      partition,
-                      keySchemaAndValueProducer,
-                      valueSchemaAndValueProducer,
-                      sourceOffset,
-                      topicName,
-                      keyDocument,
-                      valueDoc)
-                  .map(sourceRecords::add);
-            });
-
-        if (sourceRecords.size() == maxBatchSize) {
-          LOGGER.debug(
-              "Reached '{}': {}, returning records", POLL_MAX_BATCH_SIZE_CONFIG, maxBatchSize);
-          return sourceRecords;
-        } else {
-          LOGGER.debug(
-              "Continuing to loop because sourceRecords size {} is less than maxBatchSize {}",
-              sourceRecords.size(),
-              maxBatchSize);
-        }
-      }
-    }
-    LOGGER.debug("Returning null because connector is no longer running");
-    return null;
-  }
-
-  private Optional<SourceRecord> createSourceRecord(
-      final Map<String, Object> partition,
-      final SchemaAndValueProducer keySchemaAndValueProducer,
-      final SchemaAndValueProducer valueSchemaAndValueProducer,
-      final Map<String, String> sourceOffset,
-      final String topicName,
-      final BsonDocument keyDocument,
-      final BsonDocument valueDocument) {
-
-    try {
-      SchemaAndValue keySchemaAndValue = keySchemaAndValueProducer.get(keyDocument);
-      SchemaAndValue valueSchemaAndValue = valueSchemaAndValueProducer.get(valueDocument);
-      return Optional.of(
-          new SourceRecord(
-              partition,
-              sourceOffset,
-              topicName,
-              keySchemaAndValue.schema(),
-              keySchemaAndValue.value(),
-              valueSchemaAndValue.schema(),
-              valueSchemaAndValue.value()));
-    } catch (Exception e) {
-      Supplier<String> errorMessage =
-          () ->
-              format(
-                  "%s : Exception creating Source record for: Key=%s Value=%s",
-                  e.getMessage(), keyDocument.toJson(), valueDocument.toJson());
-      if (sourceConfig.logErrors()) {
-        LOGGER.error(errorMessage.get(), e);
-      }
-      if (sourceConfig.tolerateErrors()) {
-        if (sourceConfig.getDlqTopic().isEmpty()) {
-          return Optional.empty();
-        }
-        return Optional.of(
-            new SourceRecord(
-                partition,
-                sourceOffset,
-                sourceConfig.getDlqTopic(),
-                Schema.STRING_SCHEMA,
-                keyDocument.toJson(),
-                Schema.STRING_SCHEMA,
-                valueDocument.toJson()));
-      }
-      throw new DataException(errorMessage.get(), e);
-    }
+    return startedTask.poll();
   }
 
   @Override
-  @SuppressWarnings("try")
-  public synchronized void stop() {
```

Comment:
Removed `synchronized` as `stop` cannot be called concurrently with itself, only with `poll`/`commit`, but `poll`/`commit` were neither `synchronized` nor used `synchronized (this)` blocks.

Reply:
Ack